flatgfa/src/flatbed.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    pub fn get_intersects(&self, bed: &FlatBED, entry: &BEDEntry) -> Vec<BEDEntry> {
+        self.entries.items()
+            .map(|x| BEDEntry {
+                name: x.1.name,
+                start: if x.1.start < entry.start { entry.start } else { x.1.start },
+                end: if entry.end < x.1.end { entry.end } else { x.1.end },
+            })
+            .filter(|x| bed.get_name_of_entry(entry).eq(self.get_name_of_entry(x)) && x.end > x.start)
+            .collect()
+    }
+}
+
+
+
+
+
+
+
+
+
+
+
+/// The data storage pools for a `FlatBED`.
+#[derive(Default)]
+pub struct BEDStore<'a, P: StoreFamily<'a>> {
+    pub name_data: P::Store<u8>,
+    pub entries: P::Store<BEDEntry>,
+}
+
+impl<'a, P: StoreFamily<'a>> BEDStore<'a, P> {
+    pub fn add_entry(&mut self, name: &[u8], start: u64, end: u64) -> Id<BEDEntry> {
+        let name = self.name_data.add_slice(name);
+        self.entries.add(BEDEntry {
+            name,
+            start,
+            end,
+        })
+    }
+
+    pub fn as_ref(&self) -> FlatBED {
+        FlatBED {
+            name_data: self.name_data.as_ref(),
+            entries: self.entries.as_ref(),
+        }
+    }
+}
+
+pub trait StoreFamily<'a> {
+    type Store<T: Clone + 'a>: Store<T>;
+}
+
+#[derive(Default)]
+pub struct HeapFamily;
+impl<'a> StoreFamily<'a> for HeapFamily {
+    type Store<T: Clone + 'a> = HeapStore<T>;
+}
+
+pub struct FixedFamily;
+impl<'a> StoreFamily<'a> for FixedFamily {
+    type Store<T: Clone + 'a> = FixedStore<'a, T>;
+}
+
+/// A store for `FlatBED` data backed by fixed-size slices.
+///
+/// This store contains `SliceVec`s, which act like `Vec`s but are allocated within
+/// a fixed region. This means they have a maximum size, but they can directly map
+/// onto the contents of a file.
+pub type FixedBEDStore<'a> = BEDStore<'a, FixedFamily>;
+
+/// A mutable, in-memory data store for `FlatBED`.
+///
+/// This store contains a bunch of `Vec`s: one per array required to implement a
+/// `FlatBED`. It exposes an API for building up a BED data structure, so it is
+/// useful for creating new ones from scratch.
+pub type HeapBEDStore = BEDStore<'static, HeapFamily>;
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+type ParseResult<T> = Result<T, &'static str>;
+type PartialParseResult<'a, T> = ParseResult<(T, &'a [u8])>;
+fn parse_num<T: FromRadix10>(s: &[u8]) -> PartialParseResult<T> {
+    match T::from_radix_10(s) {
+        (_, 0) => Err("expected number"),
+        (num, used) => Ok((num, &s[used..])),
+    }
+}
+
+pub struct BEDParser<'a, P: StoreFamily<'a>> {
+    /// The flat representation we're building.
+    flat: BEDStore<'a, P>,
+}
+
+impl<'a, P: StoreFamily<'a>> BEDParser<'a, P> {
+    pub fn new(builder: BEDStore<'a, P>) -> Self {
+        Self {
+            flat: builder,
+        }
+    }
+
+    /// Parse a GFA text file from an I/O stream.
+    pub fn parse_stream<R: BufRead>(mut self, stream: R) -> BEDStore<'a, P> {
```

Comment:
Maybe we can delete this? Not sure we really need to parse streams… I know I did this for GFA files, but it's really just kinda for show and probably not important because `parse_mem` is likely to be so much faster.